scripts/install_xcpng.py

```diff
@@ -0,0 +1,225 @@
+        ssh(pxe, ['rm', '-rf', distant_directory])
+    else:
+        raise Exception('ERROR: mac address is null or empty!')
+
+def clean_bootconf_on_pxe(mac_address):
+    pxe = pxe_address()
+    if mac_address:
+        distant_file = f'/pxe/configs/custom/{mac_address}/boot.conf'
+        try:
+            ssh(pxe, ['rm', '-rf', distant_file])
+        except SSHCommandFailed:
+            raise Exception('ERROR: problem to clean the boot.conf file.')
+    else:
+        raise Exception('ERROR: mac address is null or empty!')
+
+def get_candidate_ips(mac_address):
+    pxe = pxe_address()
+    ip_address = ssh(
+        pxe,
+        ['arp', '-n', '|', 'grep', mac_address, '|', 'awk', '\'{ print $1 }\'']
+    )
+    candidate_ips = ip_address.splitlines()
+    return candidate_ips
+
+def is_ip_active(ip):
+    ret = os.system(f"ping -c 3 -W 10 {ip} > /dev/null 2>&1")
+    if ret != 0:
+        return False
+    else:
+        return True
+
+def is_ssh_up(ip):
+    try:
+        ssh(ip, ['true'])
+        return True
+    except SSHCommandFailed:
+        # probably not up yet
+        return False
+
+def get_new_host_ip(mac_address):
+    candidate_ips = get_candidate_ips(mac_address)
+    for ip in candidate_ips:
+        if is_ip_active(ip) and is_ssh_up(ip):
+            return ip
+    return None
+
+def is_new_host_ready(ip_address):
+    try:
+        output = ssh(ip_address, ['xe', 'host-list', 'enabled=true', '--minimal']).strip()
+        return is_uuid(output)
+    except Exception:
+        return False
+
+def check_mac_address(host, mac_address):
+    bridge = host.inventory['MANAGEMENT_INTERFACE']
+    host_mac_address = host.ssh(['cat', f'/sys/class/net/{bridge}/address'])
+    if mac_address != host_mac_address:
+        raise Exception(
+            f"Unexpected MAC address {host_mac_address} for host {host.hostname_or_ip}. Expected: {mac_address}"
+        )
+
+def url_checker(url):
+    try:
+        response = requests.get(url)
+        if not response:
+            raise Exception(f"{url}: URL is not reachable, status_code: {response.status_code}")
+    except requests.exceptions.RequestException as e:
+        raise SystemExit(f"{url}: URL is not reachable\nErr: {e}")
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("host", help="The IP address or the name of the host where the VM is created.")
+    parser.add_argument("vm_uuid", help="The UUID of an existing VM to use to create the new host.")
+    parser.add_argument("action", metavar='action', choices=['install', 'upgrade'], help="install or upgrade.")
+    parser.add_argument("version_xcpng", help="The version of xcp-ng we wish.")
+    parser.add_argument("--installer", help="URL of the installer on the PXE")
+    args = parser.parse_args()
+
+    try:
+        is_uuid(args.vm_uuid)
+    except Exception:
+        raise Exception(f'The VM UUID given is invalid: {args.vm_uuid}')
```

Comment:
Display exception?

Reply:
The logic is wrong. is_uuid returns a boolean.

Reply:
indeed
```py
if not is_uuid(args.vm_uuid):
    raise Exception(f'The VM UUID given is invalid: {args.vm_uuid}')
```

Reply:
I think this has not been addressed and/or answered.

Reply:
Right, @gduperrey must have overlooked it. Done locally.